producer_sarama.go

```diff
@@ -0,0 +1,183 @@
+	case "none":
+		config.Compression = sarama.CompressionNone
+	case "gzip":
+		config.Compression = sarama.CompressionGZIP
+	case "snappy":
+		config.Compression = sarama.CompressionSnappy
+	}
+	config.FlushByteCount = conf.FlushByteCount
+	config.FlushFrequency = conf.FlushTimeout
+	config.FlushMsgCount = conf.BatchSize
+	config.MaxMessageBytes = conf.MaxMessageBytes
+	config.MaxMessagesPerReq = conf.MaxMessagesPerRequest
+	config.RequiredAcks = sarama.RequiredAcks(conf.Acks)
+	config.RetryBackoff = conf.RetryBackoff
+	config.Timeout = conf.Timeout
+
+	partitionerFactory := &SaramaPartitionerFactory{conf.Partitioner}
+	config.Partitioner = partitionerFactory.PartitionerConstructor
+
+	producer, err := sarama.NewProducer(client, config)
+	if err != nil {
+		panic(err)
+	}
+	return &SaramaProducer{
+		saramaProducer: producer,
+	}
+}
+
+func (this *SaramaProducer) Errors() <-chan *FailedMessage {
+	errorChannel := make(chan *FailedMessage)
+	go func() {
+		for saramaError := range this.saramaProducer.Errors() {
+			key, err := saramaError.Msg.Key.Encode()
+			if err != nil {
+				panic(err)
+			}
+			value, err := saramaError.Msg.Value.Encode()
+			if err != nil {
+				panic(err)
+			}
+			msg := &ProducerMessage{
+				Topic:     saramaError.Msg.Topic,
+				Key:       key,
+				Value:     value,
+				partition: saramaError.Msg.Partition(),
+				offset:    saramaError.Msg.Offset(),
+			}
+			errorChannel <- &FailedMessage{msg, saramaError.Err}
+		}
+	}()
+
+	return errorChannel
+}
+
+func (this *SaramaProducer) Successes() <-chan *ProducerMessage {
+	successChannel := make(chan *ProducerMessage)
+	go func() {
+		for saramaMessage := range this.saramaProducer.Successes() {
+			key, err := saramaMessage.Key.Encode()
+			if err != nil {
+				panic(err)
+			}
+			value, err := saramaMessage.Value.Encode()
+			if err != nil {
+				panic(err)
+			}
+			msg := &ProducerMessage{
+				Topic:     saramaMessage.Topic,
+				Key:       key,
+				Value:     value,
+				partition: saramaMessage.Partition(),
+				offset:    saramaMessage.Offset(),
+			}
+			successChannel <- msg
+		}
+	}()
+
+	return successChannel
+}
+
+func (this *SaramaProducer) Input() chan<- *ProducerMessage {
+	messageChannel := make(chan *ProducerMessage)
+	go func() {
+		for message := range messageChannel {
+			encodedKey, err := message.KeyEncoder.Encode(message.Key)
+			if err != nil {
+				panic(err)
+			}
+			key := sarama.ByteEncoder(encodedKey)
+
+			encodedValue, err := message.ValueEncoder.Encode(message.Value)
+			if err != nil {
+				panic(err)
+			}
+			value := sarama.ByteEncoder(encodedValue)
+			saramaMessage := &sarama.ProducerMessage{
+				Topic: message.Topic,
+				Key:   key,
+				Value: value,
+			}
+			this.saramaProducer.Input() <- saramaMessage
+		}
+	}()
+
+	return messageChannel
+}
+
+func (this *SaramaProducer) Close() error {
+	return this.saramaProducer.Close()
+}
+
+func (this *SaramaProducer) AsyncClose() {
+	this.saramaProducer.AsyncClose()
```

Comment:
I think it would be cool not to use sarama's AsyncClose but use our async close style thing returning a channel